File: 13_回溯算法/经典bt/267. 回文排列 II.ts
/**
 * @param {string} s
 * @return {string[]}
 * @description
 * 返回其通过重新排列组合后所有可能的回文字符串，并去除重复的组合。
 * @summary
 * 根据字符的个数构造回文串即可，每次在首尾两个位置看能否填充两个相同的字符(如果是字符串长是奇数中间位置特判一下)
 */
function generatePalindromes(s: string): string[] {
  const res: string[] = []
  const counter = new Map<string, number>()
  for (const char of s) {
    counter.set(char, (counter.get(char) || 0) + 1)
  }

  dfs(0, s.length - 1, Array(s.length).fill(''))
  return res

  // 因为直接覆盖path 所以回溯不需要重置 path[left] 和 path[right]
  function dfs(left: number, right: number, path: string[]): void {
    if (left > right) {
      res.push(path.join(''))
      return
    }

    for (const [char, frequency] of counter.entries()) {
      if (frequency < 1) continue
      if (left !== right && frequency < 2) continue

      counter.set(char, counter.get(char)! - (left < right ? 2 : 1))
      path[left] = char
      path[right] = char
      dfs(left + 1, right - 1, path)
      counter.set(char, counter.get(char)! + (left < right ? 2 : 1))
    }
  }
}

console.log(generatePalindromes('aabb'))
// 输出: ["abba", "baab"]

File: 13_回溯算法/经典bt/267. 回文排列 II.py
# 267. 回文排列 II
# 全排列生成回文的一半
# 1 <= s.length <= 16
# s 仅由小写英文字母组成

from collections import Counter
from itertools import permutations
from typing import List


class Solution:
    def generatePalindromes(self, s: str) -> List[str]:
        counter = Counter(s)
        odd, even = [], []
        for k, v in counter.items():
            if v & 1:
                odd.append(k)
            even.extend([k] * (v // 2))

        if len(odd) > 1:
            return []

        res = set()
        for perm in permutations(even):
            res.add("".join(perm) + "".join(odd) + "".join(perm[::-1]))
        return list(res)


print(Solution().generatePalindromes("aabb"))
